src/Raven.Server/Utils/ShardHelper.cs

```diff
@@ -194,11 +162,82 @@ public static IEnumerable<string> GetShardNames(string databaseName, int shardsC
+            return FindBucketShard(configuration.BucketRanges, bucket);
+        }
+
+        public static int GetShardNumberFor(ShardingConfiguration configuration, ByteStringContext allocator, string id)
+        {
+            using (DocumentIdWorker.GetLower(allocator, id, out var lowerId))
+            {
+                return GetShardNumberFor(configuration, lowerId);
+            }
+        }
+
+        public static int GetShardNumberFor<TTransaction>(ShardingConfiguration configuration, TransactionOperationContext<TTransaction> context, string id)
+            where TTransaction : RavenTransaction
+        {
+            return GetShardNumberFor(configuration, context.Allocator, id);
+        }
+
+        public static int GetShardNumberFor(ShardingConfiguration configuration, ByteStringContext allocator, LazyStringValue id)
+        {
+            DevelopmentHelper.ShardingToDo(DevelopmentHelper.TeamMember.Arek, DevelopmentHelper.Severity.Normal, "Avoid the allocation of the LazyStringValue below");
+            using (DocumentIdWorker.GetLower(allocator, id, out var lowerId))
+            {
+                return GetShardNumberFor(configuration, lowerId);
+            }
+        }
+
+        public static int GetShardNumberFor<TTransaction>(RawShardingConfiguration configuration, TransactionOperationContext<TTransaction> context, string id)
+            where TTransaction : RavenTransaction
+        {
+            int bucket = GetBucketFor(context, id);
+            int index = 0;
+            foreach (var (prefix, _) in configuration.Prefixed)
+            {
+                index++;
+                if (id.StartsWith(prefix, StringComparison.OrdinalIgnoreCase))
+                {
+                    bucket += index << 20;
+                    break;
+                }
+            }
+
+            return FindBucketShard(configuration.BucketRanges, bucket);
+        }
+
+        public static int GetShardNumberFor(ShardingConfiguration configuration, Slice id)
+        {
+            int bucket = GetBucketFromSlice(id);
+            int index = 0;
+            foreach (var (prefix, _) in configuration.Prefixed)
+            {
+                index++;
+
+                if (id.ToString().StartsWith(prefix, StringComparison.OrdinalIgnoreCase))
+                {
+                    bucket += index << 20;
+                    break;
+                }
+            }
+
+            return FindBucketShard(configuration.BucketRanges, bucket);
+        }
+
+        private static int FindBucketShard(List<ShardBucketRange> ranges, int bucket)
+        {
+            int prefixRange = bucket >> 20;
             for (int i = 0; i < ranges.Count - 1; i++)
             {
-                if (bucket < ranges[i + 1].BucketRangeStart)
+                int bucketRangeStart = ranges[i].BucketRangeStart;
+                if ((bucketRangeStart >> 20) != prefixRange)
```

Comment:
This approach limits us to ~12 ranges, no?
But if we use additions it will have ~4K

Reply:
yes, we should be able to support up to 4k prefixes

Reply:
fixed